plutus-core/plutus-ir/test/GeneratorSpec.hs

```diff
@@ -0,0 +1,271 @@
+--             & set UPLC.soInlineHints hints
+--
+--   plcT <- flip runReaderT pirCtx $ PIR.compileReadableToPlc $ fmap Original tm
+--   plcTcError <- runExceptT @(PLC.Error _ _ _)
+--              $ UPLC.deBruijnTerm =<< UPLC.simplifyTerm uplcSimplOpts (UPLC.erase plcT)
+--   case plcTcError of
+--     Left _   -> error "wrong"
+--     Right cc -> return $ DeserializedCode (UPLC.Program () (PLC.defaultVersion ()) $ void cc) Nothing mempty
+--
+-- prop_compile :: Property
+-- prop_compile =
+--   forAllDoc "_,tm" genTypeAndTermNoHelp_ (shrinkTypedTerm mempty mempty) $ \ (_, tm) ->
+--   isRight $ compile tm
+
+generators :: TestNested
+generators = return $ testGroup "generators"
+  [ testProperty "prop_genKindCorrect"  $ withMaxSuccess 1000 prop_genKindCorrect
+  , testProperty "prop_shrinkTypeSound" $ prop_shrinkTypeSound
+  , testProperty "prop_genTypeCorrect"  $ withMaxSuccess 1000 prop_genTypeCorrect
+  , testProperty "prop_shrinkTermSound" $ withMaxSuccess 20 prop_shrinkTermSound
+  ]
+
+-- * Core properties for PIR generators
+
+-- | Check that the types we generate are kind-correct
+prop_genKindCorrect :: Property
+prop_genKindCorrect =
+  forAllDoc "ctx" genCtx (const []) $ \ ctx ->
+  forAllDoc "k,ty" genKindAndType (shrinkKindAndType ctx) $ \ (k, ty) ->
+  checkKind ctx ty k
+
+-- | Check that shrinking types maintains kinds
+prop_shrinkTypeSound :: Property
+prop_shrinkTypeSound =
+  forAllDoc "k,ty" genKindAndType (shrinkKindAndType Map.empty) $ \ (k, ty) ->
+  -- See discussion about the same trick in `prop_shrinkTermSound`
+  checkKind Map.empty ty k ==>
+  checkNoCounterexamples [ (k, ty) | (k, ty) <- shrinkKindAndType Map.empty (k, ty)
+                                   , not $ checkKind Map.empty ty k ]
+
+-- | Test that our generators only result in well-typed terms.
+prop_genTypeCorrect :: Property
+prop_genTypeCorrect =
+  forAllDoc "ty,tm" genTypeAndTerm_ (const []) $ \ (ty, tm) -> typeCheckTerm tm ty
+
+-- | Test that when we generate a fully applied term we end up
+-- with a well-typed term.
+prop_genWellTypedFullyApplied :: Property
+prop_genWellTypedFullyApplied =
+  forAllDoc "ty, tm" genTypeAndTerm_ (shrinkTypedTerm mempty mempty) $ \ (ty, tm) ->
+  forAllDoc "ty', tm'" (genFullyApplied ty tm) (const []) $ \ (ty', tm') -> typeCheckTerm tm' ty'
+
+-- | Test that shrinking a well-typed term results in a well-typed term
+prop_shrinkTermSound :: Property
+prop_shrinkTermSound =
+  forAllShrinkBlind (pure False) (\ sh -> [ True | not sh ]) $ \ _ ->
+  forAllDoc "ty,tm"   genTypeAndTerm_ shrinkClosedTypedTerm $ \ (ty, tm) ->
+  let shrinks = shrinkClosedTypedTerm (ty, tm) in
+  -- While we generate well-typed terms we still need this check here for
+  -- shrinking counterexamples to *this* property. If we find a term whose
+  -- shrinks aren't well-typed we want to find smaller *well-typed* terms
+  -- whose shrinks aren't well typed.
```

Comment:
Super clever.

I'm confused about something, though. So let's pretend those pairs are numbers and "type checking" a number fails when that number is, say, 17.

We start running tests:

- 0: no shrinks, skip
- 1: all shrinks are well-typed
- 3: all shrinks are well-typed
- ...
- 25: one of the shrinks is ill-typed, 17

Now we start shrinking 25.

- 0: all shrinks are well-typed
- 2: all shrinks are well-typed 
- 4: all shrinks are well-typed
- ...
- 17: uh-oh, ill-typed itself, skip
- 20: found a smaller number that shrinks to an ill-typed one

Now reduce 20:

- 0: all shrinks are well-typed
- ...

So in total: any counterexample will require us to go through a cubic (?) number of boring test cases. Is this correct?

Of course type-and-term pairs are not numbers, so we aren't going to test the same thing over and over again, but it does seem like we're mostly stuck in the corner with little examples.

Or am I misunderstanding it? \

Reply:
I unresolved this comment, because I still have the questions.

Reply:
Oops, I don't think this was meant to get resolved. I was intending to answer here. Yes, you're right that - in the worst case - you get cubic shrinking. There are two things you can do about this:
1. Ignore it! You're about to report a bug to the developer - they can afford to wait!
2. Do something clever about the shrinking order to avoid this worst-case as often as possible. There are various things you can do - like trying to shrinks in reverse order etc. But it's not clear what the right heuristic is in this case.

Reply:
Yes, the user sees that shrinking is taking place. The reason I hesitate to put a comment in about this is that this is almost always the case for all QuickCheck properties - not just this one.